src/catchup/IndexBucketsWork.cpp

```diff
@@ -0,0 +1,88 @@
+// Copyright 2022 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "IndexBucketsWork.h"
+#include "bucket/BucketIndex.h"
+#include "bucket/BucketList.h"
+#include "bucket/BucketManager.h"
+#include "util/HashOfHash.h"
+#include "util/types.h"
+#include "work/WorkWithCallback.h"
+#include <Tracy.hpp>
+
+namespace stellar
+{
+IndexBucketsWork::IndexBucketsWork(Application& app)
+    : Work(app, "index-bucketList", BasicWork::RETRY_NEVER)
+    , mBM(app.getBucketManager())
+{
+}
+
+BasicWork::State
+IndexBucketsWork::doWork()
+{
+    ZoneScoped;
+    if (!mChildrenSpawned)
+    {
+        auto indexBucket = [&](std::shared_ptr<Bucket> b) {
+            // Don't index empty bucket or buckets that are already being
+            // indexed. Sometimes one level's snap bucket may be another level's
+            // future bucket. The indexing job may have started but not
+            // finished, so check mIndexedBuckets. Some buckets may have already
+            // been indexed when assuming LCL state, so also check b->isIndexed.
+            if (b->getFilename().empty() || b->isIndexed() ||
+                mIndexedBuckets.find(b->getHash()) != mIndexedBuckets.end())
+            {
+                return;
+            }
+
+            mIndexedBuckets.insert(b->getHash());
+            auto indexCallback = [b](Application& app) {
+                b->setIndex(BucketIndex::createIndex(app.getConfig(),
+                                                     b->getFilename()));
+                return true;
+            };
+
+            addWork<WorkWithCallback>("index-" + hexAbbrev(b->getHash()),
```

Comment:
This is still problematic because it's done from the main thread, so this will cause nodes to go out of sync

Reply:
I think this code path is only for startup. When a validator is running, the merging is done on the background thread in `Bucket::merge` and synchronously for small buckets in `Bucket::fresh`. `IndexBucketsWork` is called during catchup and when loading the LCL initially, so it should be ok to block here?

Reply:
We can't block during online catchup because we need to continue listening to the network to buffer ledgers

Reply:
Is this why the apply buckets work has the `advance` interface to break it down into fast chunks to not block and should we use a similar interface here?

Reply:
yeah, that's why bucket application is done in chunks. Regarding what to do here, I don't think we need to mimic that pattern as it's super slow and requires a bunch of changes to how we index: simplest is to just spin up indexing in the background, and notify Work when it's ready. We have this pattern in a few places, see VerifyBucketWork for example (it sends an expensive verification task to the background, then the background task schedules a callback to the main thread to notify the work it's ready - it's not the cleanest pattern tbh, but it's sufficient for now)